<launch>

    <arg name="enable_debug" default="false"/>
    <!-- <arg name="enable_ctm" default="true"/> -->
    <arg name="enable_octomap" default="true"/>
    <arg name="enable_marble_mapping" default="false"/>
    <arg name="enable_inner_crop" default="true"/>
    <arg name="enable_outer_crop" default="false"/>

    <arg name="fixed_frame_id" default="map"/>
    <arg name="child_frame_id" default="base_link"/>
    <arg name="input_cloud_topic" default="points"/>

    <arg name="octomap_resolution" default="0.2"/>
    <arg name="range" default="20.0"/>
    <arg name="trav_marker_density"             default="0"/>
    <arg name="trav_marker_lower_thresh"        default="0.0"/>
    <arg name="trav_marker_upper_thresh"        default="1.0"/>

    <arg name="enable_traversability"           default="false" />
    <arg name="enable_traversability_sharing"   default="false" />

    <!-- OCTOMAP -->
    <group >
        <include file="$(find octomap_server)/launch/octomap_mapping_rough.launch">
            <!-- <arg name="ns"             value="$(arg ns)"/> -->
            <arg name="cloud_in"                value="$(arg input_cloud_topic)/cost_norm_cloud/cropped2"/>
            <arg name="frame_id"                value="$(arg fixed_frame_id)"/>
            <arg name="base_frame_id"           value="$(arg child_frame_id)"/>
            <arg name="filter_speckles"         value="false"/>
            <arg name="filter_ground"           value="false"/>
            <arg name="sensor_model/max_range"  value="$(arg range)"/>
            <arg name="trav_marker_density"     value="$(arg trav_marker_density)"/>
            <arg name="trav_marker_lower_thresh"        value="$(arg trav_marker_lower_thresh)"/>
            <arg name="trav_marker_upper_thresh"        value="$(arg trav_marker_upper_thresh)"/>
            <arg name="resolution"              value="$(arg octomap_resolution)"/>
            <arg name="sensor_model/hit"        value="0.8"/>
            <arg name="sensor_model/miss"       value="0.45"/>
            <arg name="sensor_model/min"        value="0.12"/>
            <arg name="sensor_model/max"        value="0.97"/>
            <arg name="compress_map"            value="true"/>
        </include>
        <!-- CURVY TERRAIN MAPPER -->
        <include file="$(find curvy_terrain_mapper)/launch/curvy_terrain_mapper.launch" >
            <arg name="fixed_frame_id"                                  value="$(arg child_frame_id)"/>
            <arg name="input_cloud_topic"                               value="$(arg input_cloud_topic)"/>
            <arg name="main_cloud_topic"                                value="$(arg input_cloud_topic)/main_cloud"/>
            <arg name="cost_cloud_topic"                                value="$(arg input_cloud_topic)/cost_cloud"/>
            <arg name="cost_norm_cloud_topic"                           value="$(arg input_cloud_topic)/cost_norm_cloud"/>
            <arg name="enable_debug"                                    value="$(arg enable_debug)"/>
            <arg name="preanalysis/dsFlag"                              value="true"/>
            <arg name="preanalysis/dsResolution"                        value="0.05"/>
            <arg name="preanalysis/gpFlag"                              value="false"/>
            <arg name="preanalysis/pfActive"                            value="false"/>
            <arg name="preanalysis/fsActive"                            value="false"/>
            <arg name="preanalysis/dsMethod"                            value="false"/>
            <arg name="costmap/normal_gain"                             value="4.0"/>
            <arg name="costmap/curv_gain"                               value="15.0"/>
            <arg name="costmap/min_saturation_cost"                     value="0.0"/>
            <arg name="costmap/max_saturation_cost"                     value="0.1"/>
            <arg name="costmap/set_min_saturation_cost_to_min_cost"     value="false"/>
            <arg name="costmap/set_max_saturation_cost_to_max_cost"     value="false"/>
        </include>
    </group>

    <!-- INNER POINTCLOUD CROP -->
    <node pkg="nodelet" type="nodelet" name="inner_crop_filter" args="standalone pcl/CropBox" output="screen" if="$(arg enable_inner_crop)">
      <remap from="~input" to="$(arg input_cloud_topic)/cost_norm_cloud" />
      <remap from="~output" to="$(arg input_cloud_topic)/cost_norm_cloud/cropped" />
      <rosparam>
        # true: no points in the box, false: no points outside the box
        negative: true
        min_x: -1.0
        max_x: 1.0
        min_y: -1.0
        max_y: 1.0
        min_z: -0.5
        max_z: 1.0
      </rosparam>
    </node>
    <node pkg="topic_tools" type="relay" name="inner_crop_relay" args="$(arg input_cloud_topic)/cost_norm_cloud $(arg input_cloud_topic)/cost_norm_cloud/cropped" unless="$(arg enable_inner_crop)"/>

    <!-- OUTER POINTCLOUD CROP -->
    <node pkg="nodelet" type="nodelet" name="outer_crop_filter" args="standalone pcl/CropBox" output="screen" if="$(arg enable_outer_crop)">
      <remap from="~input" to="$(arg input_cloud_topic)/cost_norm_cloud/cropped" />
      <remap from="~output" to="$(arg input_cloud_topic)/cost_norm_cloud/cropped2" />
      <rosparam>
        # true: no points in the box, false: no points outside the box
        negative: false
        min_x: -1.0
        max_x: 10.0
        min_y: -5.0
        max_y: 5.0
        min_z: -0.25
        max_z: 0.25
      </rosparam>
    </node>
    <node pkg="topic_tools" type="relay" name="outer_crop_relay" args="$(arg input_cloud_topic)/cost_norm_cloud/cropped $(arg input_cloud_topic)/cost_norm_cloud/cropped2" unless="$(arg enable_outer_crop)"/>

</launch>
